exchanges/graphcache/src/store/data.ts

```diff
@@ -527,36 +525,34 @@ export const reserveLayer = (
   layerKey: number,
   hasNext?: boolean
 ) => {
+  // Find the current index for the layer, and remove it from
+  // the order if it exists already
+  let index = data.optimisticOrder.indexOf(layerKey);
+  if (index > -1) data.optimisticOrder.splice(index, 1);
+
   if (hasNext) {
     data.deferredKeys.add(layerKey);
+    // If the layer has future results then we'll move it past any layer that's
+    // still empty, so currently pending operations will take precedence over it
+    for (
+      index = index > -1 ? index : 0;
```

Comment:
This is the relevant changed line. We shift the layer based on where it was before if it's still around, rather than letting it "hop back in front"